pkg/icingadb/history/retention.go

```diff
@@ -180,9 +181,14 @@ func (r *Retention) Start(ctx context.Context) error {
 			zap.Uint64("retention-days", days),
 		)
 
+		if days > -math.MinInt {
```

Comment:
What if we switch from `uint64` to `uint32`?

Reply:
The (non) issue is that the size of `int` dependents on your platform, either being 32 or 64 bits. In addition, the min values is always one less in the negative direction than the max value in the positive (-9223372036854775808 and 9223372036854775807 for `int` on my amd64).

Thus, there may be the case that days is within `(-math.MinInt, math.MaxUint64]`, which this [playground example](https://go.dev/play/p/YF8TiQPPZoN) should illustrate.

However, this is not a real issue when being put into context. The value of `days` will be used for the retention and if a user decides to overflow (or is it underflow now?) `math.MinInt`, we are dealing with over `2.526951242973911e+16` years. Thus, this is just a sanity check and may prevent us from potential crashes if a user decides to put a fantasy-grade number there.

Reply:
Are the `int` bounds actually the relevant thing here? Like is the `time` package even able to represent the result of `tick.Time.AddDate(0, 0, -int(days))` with `days` being $2^{31}-1$ or even $2^{63}-1$?

Reply:
Not really as `time.Time` may overflow already. 

The following snippet demonstrates this:
```go
	t := time.Now()
	fmt.Printf("%v\n%v\n", t, t.AddDate(0, 0, -math.MaxUint32))
```
Resulting in jumping back one day:
```
2024-10-08 10:32:27.911974323 +0200 CEST m=+0.000032968
2024-10-09 10:32:27.911974323 +0200 CEST
```

This is known behavior, as shown in https://github.com/golang/go/issues/20678 and other issues.

But even adding "just" `math.MaxUint32` does already overflow the time. Thus, I think there is not much we can do here and if a user decides to write astrophysical space time numbers, than let it be.

I moved the check up and decide to bail out of the retention if the user has configured more than $2^{16}$ days. This is totally arbitrary, I know, but may safe a setup from purging new data, as shown in my example above.

Reply:
> I moved the check up and decide to bail out of the retention if the user has configured more than $2^{16}$ days. This is totally arbitrary, I know, but may safe a setup from purging new data, as shown in my example above.

Wouldn't it make sense to just change the type in the config structs to `uint16` instead? Then it would directly reject the (probably wrong/unintended) config instead of just doing nothing with a warning. Also, that's about 180 years, that sounds like more then enough for regular use and 180 years in the past should be fine for the `time` package, so wouldn't that solve all overflow issues? Maybe this will even work without any linter comments.

Reply:
> and 180 years in the past should be fine for the `time` package, so wouldn't that solve all overflow issues?

Addendum: Probably not all. The database schema uses unsigned Unix timestamps, so anything before 1970 (which is less than 180 years ago, alternatively this would fix itself in about 126 years :see_no_evil:) would probably still overflow somewhere. `uint8`, i.e. 255 days would be too small, so maybe we want an extra check somewhere after all. Though that could just be a comparison with Unix time 0, if it's before that, just skipping the deletion (that would even result in perfectly correct behavior, there can't be anything earlier than that in the database).

Reply:
Good point. Thus I altered the if-guard to ensure that the value will neither overflow int64 nor result in a negative timestamp when being used in subtraction.

https://github.com/Icinga/icingadb/compare/1cd597b0d8887737edf033f034f366375f35efc1..f3fdc1644a79dfa12402b61de03576afdabab668